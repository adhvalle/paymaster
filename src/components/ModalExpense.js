import { useState } from 'react'

function ModalExpense ({ showModal, payments, closeModal, addExpense }) {
  if (!showModal) return false

  const [name, updateName] = useState('')
  const [amount, updateAmount] = useState(0)
  const [description, updateDescription] = useState('')

  const handleDescription = event => {
    const newDescription = event.target.value
    updateDescription(newDescription)
  }

  const handleAmount = event => {
    const newAmount = event.target.value ? parseInt(event.target.value) : event.target.value
    updateAmount(newAmount)
  }

  const handleSelectName = event => {
    const newName = event.target.value
    updateName(newName)
  }

  const handleSubmit = event => {
    event.preventDefault()
    addExpense({
      name,
      amount,
      description,
      created: Date.now()
    })
  }

  return (
    <div
      className='modal'
      onClick={closeModal}
    >
      <div
        className='modal__inner'
        onClick={e => { e.stopPropagation() }}
      >
        <form
          className='form'
          onSubmit={handleSubmit}
        >
          <div>
            <label className='form__label'>Amigo</label>
            <select onChange={handleSelectName}>
              {
                payments.map(({ name }) => (
                  <option value={name} key={name}>{name}</option>
                ))
              }
            </select>
          </div>
          <div>
            <label className='form__label'>Gasto</label>
            <input
              className='form__input'
              type='number'
              value={amount}
              onChange={handleAmount}
            />
          </div>
          <div>
            <label className='form__label'>Descripción</label>
            <input
              className='form__input'
              type='text'
              value={description}
              onChange={handleDescription}
            />
          </div>
          <button>Añadir</button>
        </form>
      </div>
    </div>
  )
}

export default ModalExpense
