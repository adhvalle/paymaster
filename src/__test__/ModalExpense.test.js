import { render, screen, act } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import ModalExpense from '../components/ModalExpense'

test('It should render the component', () => {
  const addExpense = true
  render(<ModalExpense showModal={addExpense} payments={[{ name: 'test name' }]} />)

  const app = screen.getByTestId('modal')
  expect(app).toMatchSnapshot()
})

test('It should render the component', async () => {
  Date.now = jest.fn(() => 1678812039071)
  const addExpenseFn = jest.fn()
  const addExpense = true
  render(<ModalExpense showModal={addExpense} payments={[{ name: 'test name' }]} addExpense={addExpenseFn} />)

  const option = screen.getByRole('combobox')
  const amount = screen.getByRole('spinbutton')
  const input = screen.getByRole('textbox')
  const button = screen.getByRole('button')

  act(() => {
    /* fire events that update state */
    userEvent.selectOptions(option, 'test name')
    userEvent.type(amount, '2000')
    userEvent.type(input, 'test description')
    userEvent.click(button)
  })
  expect(addExpenseFn).toHaveBeenCalledWith({
    name: 'test name',
    amount: 2000,
    description: 'test description',
    created: Date.now()
  })
})
