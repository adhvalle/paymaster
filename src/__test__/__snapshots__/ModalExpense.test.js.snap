// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render the component 1`] = `
<div
  class="modal"
  data-testid="modal"
>
  <div
    class="modal__inner"
  >
    <form
      class="form"
    >
      <div>
        <label
          class="form__label"
        >
          Amigo
        </label>
        <select>
          <option
            value=""
          >
            -
          </option>
          <option
            value="test name"
          >
            test name
          </option>
        </select>
      </div>
      <div>
        <label
          class="form__label"
        >
          Gasto
        </label>
        <input
          class="form__input"
          data-testid="number"
          type="number"
          value="0"
        />
      </div>
      <div>
        <label
          class="form__label"
        >
          Descripción
        </label>
        <input
          class="form__input"
          type="text"
          value=""
        />
      </div>
      <button>
        Añadir
      </button>
    </form>
  </div>
</div>
`;
